execution/function/function.go

```diff
@@ -1,247 +1,482 @@
+	return fmt.Sprintf("[*functionOperator] %v(%v)", o.funcExpr.Func.Name, o.funcExpr.Args), o.nextOps
+}
+
+func (o *functionOperator) Series(ctx context.Context) ([]labels.Labels, error) {
+	if err := o.loadSeries(ctx); err != nil {
+		return nil, err
+	}
+
+	return o.series, nil
+}
+
+func (o *functionOperator) GetPool() *model.VectorPool {
+	return o.nextOps[o.vectorIndex].GetPool()
+}
+
+func (o *functionOperator) Next(ctx context.Context) ([]model.StepVector, error) {
+	if err := o.loadSeries(ctx); err != nil {
+		return nil, err
+	}
+
+	// Process non-variadic single/multi-arg instant vector and scalar input functions.
+	// Call next on vector input.
+	input, err := o.nextOps[o.vectorIndex].Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(input) == 0 {
+		return nil, nil
+	}
+
+	// Call next on all scalar inputs.
+	if len(o.scalarInput) != 0 {
+		s := 0
+		for i := range o.nextOps {
+			if i == o.vectorIndex {
+				continue
+			}
+
+			si, err := o.nextOps[i].Next(ctx)
+			if err != nil {
+				return nil, err
+			}
+
+			o.scalarInput[s] = si
+			defer o.nextOps[i].GetPool().PutVectors(o.scalarInput[s])
+			s++
+		}
+	}
+
+	for v := range input {
+		// scalar() depends on number of samples per vector and returns NaN if len(samples) != 1.
+		// So need to handle this separately here, instead of going via call which is per point.
+		if o.funcExpr.Func.Name == "scalar" {
+			if len(input[v].Samples) <= 1 {
+				continue
+			}
+
+			input[v].Samples = input[v].Samples[:1]
+			input[v].SampleIDs = input[v].SampleIDs[:1]
+			input[v].Samples[0] = math.Float64frombits(value.NormalNaN)
+			continue
+		}
+
+		for i := range input[v].Samples {
+			// Scalar expressions can select from storage as well.
+			if len(o.funcExpr.Func.ArgTypes) > 1 && len(o.scalarInput) > 0 {
+				for l := range o.scalarInput {
+					if len(o.scalarInput[l]) >= 1 {
+						o.scalarPoints[l] = o.scalarInput[l][v].Samples[0]
+					}
+				}
+			}
+
+			// Call function by separately passing major input and scalars.
+			result := o.call(FunctionArgs{
+				Labels:       o.series[0],
+				Points:       []promql.Point{{V: input[v].Samples[i]}},
+				StepTime:     input[v].T,
+				ScalarPoints: o.scalarPoints,
+			})
+
+			input[v].Samples[i] = result.V
+		}
+	}
+
+	defer o.nextOps[o.vectorIndex].GetPool().PutVectors(input)
+	return input, nil
+}
+
+func (o *functionOperator) loadSeries(ctx context.Context) error {
+	var err error
+	o.once.Do(func() {
+		if o.funcExpr.Func.Name == "vector" {
+			o.series = []labels.Labels{labels.New()}
+			return
+		}
+
+		if o.funcExpr.Func.Name == "scalar" {
+			o.series = []labels.Labels{}
+			return
+		}
+
+		series, loadErr := o.nextOps[o.vectorIndex].Series(ctx)
+		if loadErr != nil {
+			err = loadErr
+		}
+
+		o.series = make([]labels.Labels, len(series))
+		for i, s := range series {
+			lbls := s
+			if o.funcExpr.Func.Name != "last_over_time" {
+				lbls = DropMetricName(s)
+			}
+
+			sort.Sort(lbls)
```

Comment:
I think labels will always be sorted, so maybe we can remove this to save some CPU cycles?